spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/MessageSerializationUtils.java

```diff
@@ -16,58 +16,48 @@
 
 package org.springframework.cloud.stream.binder;
 
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
+import java.nio.charset.StandardCharsets;
 
 import org.springframework.messaging.Message;
 import org.springframework.messaging.MessageHeaders;
-import org.springframework.messaging.converter.ContentTypeResolver;
-import org.springframework.util.MimeType;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
 
 /**
  * Utility class for serializing and de-serializing the message payload.
  *
  * @author Soby Chacko
  * @author Vinicius Carvalho
+ * @author Oleg Zhurakousky
  */
+@Deprecated
 public abstract class MessageSerializationUtils {
 
-	private static final Map<String, Class<?>> payloadTypeCache = new ConcurrentHashMap<>();
-
 	/**
 	 * Serialize the message payload unless it is a byte array.
 	 *
 	 * @param message the message with the payload to serialize
-	 * @return the Message with teh serialized payload
+	 * @return the Message with the serialized payload
 	 */
 	public static MessageValues serializePayload(Message<?> message) {
 		Object originalPayload = message.getPayload();
+		boolean setOriginalContentType = (originalPayload instanceof String);
+		Assert.isTrue(originalPayload instanceof byte[] || originalPayload instanceof String,
+				"Failed to convert message's payload. No suitable converter found for provided contentType: "
+						+ message.getHeaders().get(MessageHeaders.CONTENT_TYPE) + " and paylod: " + originalPayload);
 		Object originalContentType = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
-		MessageValues messageValues = new MessageValues(message);
-		messageValues.setPayload(originalPayload);
-		messageValues.put(MessageHeaders.CONTENT_TYPE, originalContentType);
-		return messageValues;
-	}
+		// Pass content type as String since some transport adapters will exclude
+		// CONTENT_TYPE Header otherwise
+		String contentType = setOriginalContentType ? JavaClassMimeTypeUtils.mimeTypeFromObject(originalPayload,
+				ObjectUtils.nullSafeToString(originalContentType)).toString() : originalContentType.toString() ;
```

Comment:
This throws an NPE when `originalContentType` is `null` and `setOriginalContentType` is `false`.

Reply:
This works around it... https://github.com/garyrussell/spring-cloud-stream-binder-kafka/commit/94daa9f2a3139d6e23628578c5c281c7e8474d7c